Add additional explanation and details for compose package (#13496)
WordPress/gutenberg: packages/compose/README.md

The `compose` package is a collection of handy [Higher Order Components](https://facebook.github.io/react/docs/higher-order-components.html) (HOCs) you can use to wrap your WordPress components and provide some basic features like: state, instance id, pure...

> The **compose** function is an alias to [flowRight](https://lodash.com/docs/#flowRight) from Lodash. It comes from functional programming world and allows to compose any number of functions. An example that illustrates it for two functions:

Reviewer: ```suggestion
The `compose` function is an alias to [flowRight](https://lodash.com/docs/#flowRight) from Lodash. It comes from functional programming, and allows you to compose any number of functions. You might also think of this as layering functions; `compose` will execute each function in sequence, passing the output of the function to the next function.

An example that illustrates it for two functions:
```

Reviewer: I tried to make this a little clearer for less technical WordPressers. I think my description is correct but I'd love an audit :)

Author: Updated in #13504

```js
const compose = ( f, g ) => x
    => f( g( x ) );
```

> Here's a simplified example of **compose** in use from our code, see [plugin-sidebar](https://github.com/WordPress/gutenberg/blob/master/packages/edit-post/src/components/sidebar/plugin-sidebar/index.js) for full code:

Reviewer: ```suggestion
Here's a simplified example of **compose** in use from Gutenberg's [`PluginSidebar` component](https://github.com/WordPress/gutenberg/blob/master/packages/edit-post/src/components/sidebar/plugin-sidebar/index.js):
```

Author: Updated in #13504

Using compose:

```js
const applyWithSelect = withSelect( ( select, ownProps ) => {
	return doSomething( select, ownProps);
} );
const applyWithDispatch = withDispatch( ( dispatch, ownProps ) => {
	return doSomethingElse( dispatch, ownProps );
} );

export default compose(
	withPluginContext,
	applyWithSelect,
	applyWithDispatch,
)( PluginSidebarMoreMenuItem );
```

> Equivalent to the following without compose:

Reviewer: ```suggestion
Without `compose`, the code would look like this:
```

Author: Updated in #13504

The `compose` package is a collection of handy [Higher Order Components](https://facebook.github.io/react/docs/higher-order-components.html) (HOCs) you can use to wrap your WordPress components and provide some basic features like: state, instance id, pure...

> The **compose** function is an alias to flowRight from Lodash. It comes from functional programming world and allows to compose any number of functions. An example that illustrates it for two functions:

Reviewer 2: Should it be placed under `Usage` section?

Author: I like it at the top since it is explaining what compose is, the top section is using examples to illustrate the definition, where the usage area is for specific use.

> The **compose** function is an alias to flowRight from Lodash. It comes from functional programming world and allows to compose any number of functions. An example that illustrates it for two functions:

Reviewer 2: It might be nice to link `flowRight` to https://lodash.com/docs/#flowRight.

const ComponentWithInstanceIdProp = withInstanceId( WrappedComponent );
```

> Refer to each Higher Order Component's README file for more details, see [list of components](https://github.com/WordPress/gutenberg/tree/master/packages/compose/src).

Reviewer: ```suggestion
For more details, you can refer to each Higher Order Component's README file. [Available components are located here.](https://github.com/WordPress/gutenberg/tree/master/packages/compose/src)
```

Author: Updated in #13504

const ComponentWithInstanceIdProp = withInstanceId( WrappedComponent );
```

Refer to each Higher Order Component's README file for more details, see [list of components](https://github.com/WordPress/gutenberg/tree/master/packages/compose/src).

> 

Reviewer 2: This footer should stay :)

> 

Reviewer 2: This footer should stay :)

Author: I added it back. To me it seems a bit off to have the footer, it is on the bottom of some documentation pages and not others. I don't feel all that strongly either way so if people like it, its fine.
